### ISSUE TITLE
Discuss changing default behaviour of NaN

### ISSUE THREAD

**Issue by jr200** [pola-rs/polars#4885]
Pandas has the convention to `skip_na=True` by default on [Series.max](https://pandas.pydata.org/docs/reference/api/pandas.Series.max.html), so there are more conventions... I think you can reason both ways. If you want to have the discussion to change this, please open a new issue, with more details on why you think it should be changed.

_Originally posted by @zundertj in https://github.com/pola-rs/polars/issues/2536#issuecomment-1250027153_

**Comment by jr200** [comment]
I wanted to discuss the possibility of changing the default treatment of NaNs in aggregate functions.

The behaviour I see is:

(1) Row-wise:

```
import numpy as np
import polars as pl

df = pl.DataFrame({'x': [None, float("nan"), -1, 1], 'zeros': [0.]*4})
df.with_columns([
    pl.max(['a', 'b']),
])

shape: (4, 3)
┌──────┬───────┬─────────┐
│ x    ┆ zeros ┆ row_max │
│ ---  ┆ ---   ┆ ---     │
│ f64  ┆ f64   ┆ f64     │
╞══════╪═══════╪═════════╡
│ null ┆ 0.0   ┆ 0.0     │
├╌╌╌╌╌╌┼╌╌╌╌╌╌╌┼╌╌╌╌╌╌╌╌╌┤
│ NaN  ┆ 0.0   ┆ 0.0     │
├╌╌╌╌╌╌┼╌╌╌╌╌╌╌┼╌╌╌╌╌╌╌╌╌┤
│ -1.0 ┆ 0.0   ┆ 0.0     │
├╌╌╌╌╌╌┼╌╌╌╌╌╌╌┼╌╌╌╌╌╌╌╌╌┤
│ 1.0  ┆ 0.0   ┆ 1.0     │
└──────┴───────┴─────────┘
```

(2) Column-wise:

```
df.select([ pl.col('x').max().alias('x_max') ])

shape: (1, 1)
┌───────┐
│ x_max │
│ ---   │
│ f64   │
╞═══════╡
│ 1.0   │
└───────┘
```


I expected:

`row_max = [0,0, NaN, 0.0, 1.]`
`x_max = [NaN]`


My thinking...

In polars, None/Null means 'missing value', so NaN can mean 'invalid result' as defined in the floating point spec.

For all floats x, (NaN≥x, NaN≤x, NaN>x, NaN<x, NaN=x) is False and NaN≠x is True. This means there is no well-defined ordering on any set containing NaN, and therefore the behaviour of min/max is undefined.

The options are:
- throw an error.
- (technically) return anything you want.
- choose a convention that gives a consistent behaviour.
  - e.g., treat NaN as missing (current behaviour)
  - e.g., always return NaN (behaviour in R)

It looks like, w.r.t. min/max, the convention chosen is to treat NaN as a missing value. 

Although this aligns with the default behaviour in Pandas Series.max(skip_na=True), I suggest this is because Pandas does not have a separate notion of 'missing value'. In contrast, the default behaviour of min/max in `R` (which in fairness also does not have a notion of 'missing value') is to return `NA`.

I have two points which I hope can convince you to change the default behavior.

(1) the ease of simulating both behaviours.

If the convention is: `pl.Series('x', [float("nan"), 0.]).max() == 0.0`, then I think I need to write this to simulate the `R` convention (please suggest an alternative if there is something simpler!).

```
(df
  .with_column(pl.max(['x', 0.]).alias('tmp_max'))
  .with_column(
    pl.when(pl.col('x').is_nan())
      .then(float("nan"))
      .otherwise(pl.col('tmp_max'))
      .alias('row_max_R')
    )
   .drop('tmp_max')
)

shape: (4, 3)
┌──────┬───────┬───────────┐
│ x    ┆ zeros ┆ row_max_R │
│ ---  ┆ ---   ┆ ---       │
│ f64  ┆ f64   ┆ f64       │
╞══════╪═══════╪═══════════╡
│ null ┆ 0.0   ┆ 0.0       │
├╌╌╌╌╌╌┼╌╌╌╌╌╌╌┼╌╌╌╌╌╌╌╌╌╌╌┤
│ NaN  ┆ 0.0   ┆ NaN       │
├╌╌╌╌╌╌┼╌╌╌╌╌╌╌┼╌╌╌╌╌╌╌╌╌╌╌┤
│ -1.0 ┆ 0.0   ┆ 0.0       │
├╌╌╌╌╌╌┼╌╌╌╌╌╌╌┼╌╌╌╌╌╌╌╌╌╌╌┤
│ 1.0  ┆ 0.0   ┆ 1.0       │
└──────┴───────┴───────────┘
```

On the other hand, if the convention is changed to: `pl.Series('x', [float("nan"), 0.]).max() == NaN`, then I can simulate the Pandas behaviour with `df.fill_nan(None).max()`. 

This is simpler to write, (relatively) more efficient, and it communicates the intention of the code without adding complexity.

(2) inconsistent behaviour with Polars rolling functions.

For example:

```
df = pl.DataFrame({
    'x_nans' : [0.,1,np.nan,np.nan,4,5,6],
    'x_nulls': [0.,1,None,None,4,5,6],
})

df.with_columns([
    pl.col('x_nans').rolling_max(2, min_periods=1).alias('roll_max(x_nans, 2)'),
    pl.col('x_nulls').rolling_max(2, min_periods=1).alias('roll_max(x_nulls, 2)'),
  ])

shape: (7, 4)
┌────────┬─────────┬─────────────────────┬──────────────────────┐
│ x_nans ┆ x_nulls ┆ roll_max(x_nans, 2) ┆ roll_max(x_nulls, 2) │
│ ---    ┆ ---     ┆ ---                 ┆ ---                  │
│ f64    ┆ f64     ┆ f64                 ┆ f64                  │
╞════════╪═════════╪═════════════════════╪══════════════════════╡
│ 0.0    ┆ 0.0     ┆ 0.0                 ┆ 0.0                  │
├╌╌╌╌╌╌╌╌┼╌╌╌╌╌╌╌╌╌┼╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌┼╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌┤
│ 1.0    ┆ 1.0     ┆ 1.0                 ┆ 1.0                  │
├╌╌╌╌╌╌╌╌┼╌╌╌╌╌╌╌╌╌┼╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌┼╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌┤
│ NaN    ┆ null    ┆ NaN                 ┆ 1.0                  │
├╌╌╌╌╌╌╌╌┼╌╌╌╌╌╌╌╌╌┼╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌┼╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌┤
│ NaN    ┆ null    ┆ NaN                 ┆ null                 │
├╌╌╌╌╌╌╌╌┼╌╌╌╌╌╌╌╌╌┼╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌┼╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌┤
│ 4.0    ┆ 4.0     ┆ NaN                 ┆ 4.0                  │
├╌╌╌╌╌╌╌╌┼╌╌╌╌╌╌╌╌╌┼╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌┼╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌┤
│ 5.0    ┆ 5.0     ┆ 5.0                 ┆ 5.0                  │
├╌╌╌╌╌╌╌╌┼╌╌╌╌╌╌╌╌╌┼╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌┼╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌┤
│ 6.0    ┆ 6.0     ┆ 6.0                 ┆ 6.0                  │
└────────┴─────────┴─────────────────────┴──────────────────────┘
```

By setting min_periods=1, I was trying to show NaNs are not being treated as missing values.

If we were to change the convention of the `rolling_*` methods to treat NaN as missing, it would then be (relatively) difficult and inefficient to recover the `R` behaviour with larger window-sizes.

**Comment by ritchie46** [comment]
I think we at least should make `NaN` consistent in all our functions. I will ensure we do that. That consistency should be that `null` is seen as lower than any value and that `NaN` is seen as higher than any value.

What does a `min` aggregation do in R on partially `NaN` values?

**Comment by jr200** [comment]
> What does a `min` aggregation do in R on partially `NaN` values?

```R
> min(c(NA, 0))
[1] NA
> max(c(NA, 0))
[1] NA

**Comment by jr200** [comment]
> That consistency should be that `null` is seen as lower than any value and that `NaN` is seen as higher than any value.

Is this the same as saying `NaN` == `+Inf`?  If so, I don't see the added value here because floats already have a dedicated representation for positive infinity. Then the question is: why favour positive infinity over negative infinity? I could also already replicate this behaviour easily with `df.fill_nan(float('inf'))`.

`NaN` is a completely different beast (imo), and the property that we should seek to preserve is it's unordered-ness. The  spec prescribes: "Every NaN shall compare unordered with everything, including itself (section 5.7)", and I think moving away from this definition will catch a lot of people off-guard.

**Comment by ritchie46** [comment]
After giving it some thought and looking at rusts defaults: https://doc.rust-lang.org/std/primitive.f32.html#method.min following IEEE-754 2008 I think the current behavior for our aggregations is correct, except for rolling kernels.

I think we should keep that as default behavior and add a `propagate_nans` argument.

**Comment by jr200** [comment]
I spent an amount of time reading more about the interaction of `min`/`max` and `NaN`'s today. While '`NaN` poisoning' makes much more sense to me, I more strongly agree with aligning to convention that the base-language (Rust) has chosen.

For others reading this thread, I hope its ok to link to a very good and in-depth technical discussion I found during my searches:

https://github.com/JuliaLang/julia/issues/7866

For my purposes, a `propagate_nans` setting would be greatly appreciated - will this be a per-function argument, or a `polars.Config` setting?

Thanks for taking the time to consider my request - huge fan of Polars btw :-)
